text/0143-versioning-and-stability.md

```diff
@@ -0,0 +1,128 @@
+
+**Semantic Conventions -** A schema defining the attributes which describe common concepts and operations which the signal observes. Note that unlike the API or SDK, stable conventions for all signals may be placed in the same package, as they are often useful across different signals.
+
+**Contrib –** plugins and instrumentation that make use of the API or SDK interfaces, but are not part of the core packages necessary for running OTel. The term contrib specifically refers to the plugins and instrumentation maintained by the OpenTelemetry SIG; it does not refer to third party plugins hosted elsewhere.
+
+## Signal lifecycle
+
+OpenTelemetry is structured around signals. Each signal represents a coherent, stand-alone set of functionality. Each signal follows a lifecycle.
+
+![drawing](img/0143_api_lifecycle.png)
+
+### Lifecycle stages
+
+**Experimental –** Bugs, breaking changes and performance issues may occur.
+
+**Stable –**  Stability guarantees apply, based on component tpye (API, SDK, Conventions, and Contrib). Long term dependencies may now be taken against these packages.
+
+**Deprecated –** this signal has been replaced but is still supported under LTS.
+
+**Removed -** a deprecated signal is no longer supported, and is removed.
+
+All signal components may become stable together, or one by one in the following order: API, SDK, Semantic Conventions, Contrib.
+
+### Stability
+
+Once a signal component is marked as stable, the following rules apply until the end of that signal’s existence.
+
+**API Stability -**
+No backward-incompatible changes to the API are allowed unless the major version number is incremented. All existing API calls must continue to compile and function against all future minor versions of the same major version. ABI compatibility is offered in languages which require it.
+
+**SDK Stability -**
+Public portions of the SDK (constructors, configuration, end-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break; ABI compatibility is not required.
+
+**Semantic Conventions Stability -**
+Semantic Conventions may not be removed once they are stable. New conventions may be added to replace usage of older conventions, but the older conventions are never removed, they will only be marked as deprecated in favor of the newer ones.
+
+**Contrib Stability -**
+Plugins and instrumentation are kept up to date, and are released simultaneously (or shortly after) the latest release of the API. The goal is to ensure users can update to the latest version of opentelemetry, and not be held back by the plugins that they depend on.
+
+Public portions of contrib packages (constructors, configuration, interfaces) must remain backwards compatible. ABI compatibility not required.
+
+Telemetry produced by contrib instrumentation must also remain stable and backwards compatible, to avoid breaking alerts and dashboard.
+
+### Deprecation
```

Comment:
There isn't any time-period described here between deprecation and removal, which makes deprecation somewhat moot.  I can simply replace old with new, and cut a new major version without deprecating.  If this is binding in some form (e.g. through a minimum "deprecation period"), we should spell that out.  If not, make it clearer that this is an optional, be-a-nice-community sort of thing.

Reply:
So, I don't want to arbitrarily pick a minimum deprecation period, and no one has pointed to any kind of convention or standard to follow. The current deprecation period should be considered "infinite" for the time being.

Reply:
I'm not quite following.  IIUC, an infinite deprecation period would imply we can't remove signals, even with a new major version, which doesn't align with the rest of the proposal.

Here is what I would suggest as a definition for deprecation:

"A deprecated signal is a signal which will be removed in the next major version, but is identical to a stable signal in all other respects.  Marking a signal deprecated does not imply a new major version must be cut, but **guarantees** that it will be removed if/when a new major version is cut.  All signals which are removed in a new major version must be marked deprecated in the old major version before releasing the new major version--no cherrypicking deprecations after-the-fact, and no removing signals which are not deprecated. Marking a signal deprecated is intended to smooth the transition to a new major version by encouraging users to move off of the deprecated signal prior to upgrading to the next major version.  A user who has moved off of all deprecated signals will not encounter any compilation errors when upgrading to the new major version."